include/ratelimit_manager.h

```diff
@@ -0,0 +1,201 @@
+#pragma once
+
+#include <string>
+#include <vector>
+#include <unordered_map>
+#include <mutex>
+#include <shared_mutex>
+#include "lru/lru.hpp"
+
+
+// Enum for rate limit ban durations from one minute to six hours
+enum class RateLimitBanDuration {
+    NO_BAN_BEFORE,
+    ONE_MINUTE,
+    FIVE_MINUTES,
+    TEN_MINUTES,
+    THIRTY_MINUTES,
+    ONE_HOUR,
+    THREE_HOURS,
+    SIX_HOURS,
+    PERMANENT
+};
+
+
+// Rate limit banned IP  structure with throttling_from and throttling_to fields.
+struct ratelimit_banned_ip_t {
+    std::string ip;
+    std::string api_key;
+    int64_t throttling_from;
+    int64_t throttling_to;
+};
+
+
+
+
+// Rate limit struct which stores if IP or api_keey isTracked and rate limits for one minute and one hour.
+struct ratelimit_tracker_t {
+
+    private:
+
+        friend class RateLimitManager;
+
+        inline static int64_t last_id;
+    public:
+
+    ratelimit_tracker_t() {
+        id =  ratelimit_tracker_t::last_id++;
+    }
+
+    ratelimit_tracker_t(bool is_tracked, bool is_allowed, bool is_banned_permanently, int64_t minute_rate_limit, int64_t hour_rate_limit) :
+        is_tracked(is_tracked), is_allowed(is_allowed), is_banned_permanently(is_banned_permanently), minute_rate_limit(minute_rate_limit), hour_rate_limit(hour_rate_limit){
+        id = ratelimit_tracker_t::last_id++;
+    }
+    int64_t id;
+    bool is_tracked = false;
+    bool is_allowed = false;
+    bool is_banned_permanently = false;
```

Comment:
We should just call this `is_banned` and call the temporary ban flag as `is_throttled` (I've added comment for that below as well).

Also, should we just have an `action` enum here (like the JSON counterpart) rather than allow/ban etc. being individual boolean flags.